capstone/src/test/java/com/google/step/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,156 @@
+    // Run this test twice to prove we're not leaking any state across tests
+    public void checkIfDatastoreWorks() {
+        DatastoreService datastoreService = DatastoreServiceFactory.getDatastoreService();
+        assertEquals(0, datastoreService.prepare(new Query("User")).countEntities(withLimit(10)));
+        datastoreService.put(new Entity("User"));
+        datastoreService.put(new Entity("User"));
+        assertEquals(2, datastoreService.prepare(new Query("User")).countEntities(withLimit(10)));
+    }
+
+    @Test
+    public void testCheckType1() {
+        checkIfDatastoreWorks();
+    }
+
+    @Test
+    public void testCheckType2() {
+        checkIfDatastoreWorks();
+    }
+
+    @Test
+    public void testIsLoggedIn() {
+        // Check if this works to set the user as logged in and not the admin
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(true);
+        UserService userService = UserServiceFactory.getUserService();
+        assertTrue(userService.isUserLoggedIn());
+        assertFalse(userService.isUserAdmin());
+    }
+
+    @Test
+    public void testIsNotLoggedIn() {
+        // Check if it works to show the user is not logged in
+        helper.setEnvIsAdmin(false).setEnvIsLoggedIn(false);
+        UserService userService = UserServiceFactory.getUserService();
+        assertFalse(userService.isUserLoggedIn());
+        // Can't also check if the user is an admin because the method will throw an error since the
+        // user is not even logged in.
+    }
+
+    @Test
+    public void testDoGetWithoutLogIn() throws IOException, SAXException {
+        // Testing doGet method without having the user logged in.
+        // Result: {"LoggedIn":"false","loginURL":"someURL"}
+
+        ServletRunner sr = new ServletRunner();
+        // The first parameter has to be the same as whatever goes after the '/' in the urlString for
+        // the GetMethodWebRequest so for example if the first parameter is 'restaurant' then the link
+        // would be: 'http://localhost:8080/restaurant' even though the actual servlet may be called
+        // something else.
+        sr.registerServlet("login", LoginServlet.class.getName());
+        ServletUnitClient sc = sr.newClient();
+
+        // This can throw the SAXException.
+        WebRequest request = new GetMethodWebRequest("http://localhost:8080/login");
+        WebResponse response = sc.getResponse(request);
+        assertNotNull("No response received", response);
+        assertEquals("content type", "application/json", response.getContentType());
+
+        // Turn content to JSon
+        JsonObject jsonObject = turnStringToJson(response.getText());
+
+        assertNotNull("Return Json", response.getText());
+        assertEquals("Json Loggedin", "false", jsonObject.get("LoggedIn").toString());
+
+        // Can't check for the link to be correct, but we can check that it is not null.
+        assertNotNull("Login URL", jsonObject.get("loginURL").toString());
```

Comment:
Would it not work to use a UserService to create the expected login url and check it for correctness against that? If this won't work let me know, just wondering if it's possible.

Reply:
Yeah you're right! I thought it was more randomly generated but they are actually the same. Corrected.